### ISSUE TITLE
BLT end-of-life

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa801#8181]
# Story

As a web team using BLT, we would like to begin transitioning off BLT since it has reached end-of-life.

# Additional Info

https://github.com/acquia/blt/issues/4736

>BLT will not be updated to support Drupal 11 – so, if and when you plan to upgrade to Drupal 11, Acquia’s BLT will be a blocker if it’s still in your project.

>Acquia will end support of BLT at the end of 2024 (December 31, 2024) and publicly archive the repository at that time

https://dev.acquia.com/tutorial/you-dont-need-blt-acquia-cloud

Multisite specific recommendation: https://github.com/acquia/drupal-recommended-settings. We cannot use cloud actions to replace cloud hooks.

# Questions

* Curious about our thinking re: adopting GrumPHP as part of our replacement strategy
* Curious about moving off travis (paid) in favor of github actions (free) given that we will be touching a lot of this at the same time.

# Proposed Solution

**Comment by joewhitsitt** [comment]
I think initial slices should be to move our helper functions like `umc` and `umd` out of BLT to the point where we only use BLT for deployment.

**Comment by GaryRidgway** [comment]
Grooming conversation...
- Remember us saying that we might potentially keep BLT, but seems like there are other options that are okay.
- BLT is large, perhaps postpone grooming on this?
- Should we be making a slice to probe where BLT is in our codebase to understand the scope of removing it.

**Comment by pyrello** [comment]
> https://dev.acquia.com/tutorial/you-dont-need-blt-acquia-cloud

This is useful for thinking about how to go about this switch. I was thinking that we would need to replace all the BLT commands with our own bespoke Robo implementations, so it is a relief to know we could potentially use composer scripts for some of it.

I'm noting that they don't provide a recommendation for replacing one of our most used commands `blt ds` (or `blt dsa`), so we would still need to figure out what that looks like.

We have a fair amount of customization and a whole set of custom commands (like the config split re-exporter). Even with the above instructions, I think there is a fair amount of complexity here until we identify everything that needs to be accounted for  and replaced, and how we are going to do it in each case.

I'm cool with an initial slice to switch an existing command or two over to a new implementation to learn something, but I feel like at some point we are going to need a full audit of what needs switched to really understand how much work this is.

**Comment by joewhitsitt** [comment]
When replacing blt:artifact:deploy with acli push:artifact we will most likely need to also configure `pre-acli-push-artifact` to prepare the repo (remove build files, etc.) https://drupal.slack.com/archives/C2HRNCHV4/p1742408196360569

**Comment by joewhitsitt** [comment]
This is starting to stack up... Composer is falling further behind.

https://github.com/uiowa/uiowa/pull/9573#issuecomment-3849875695

https://github.com/uiowa/uiowa/pull/9762/changes/7532734c5b04dc9ee794d942b6cb37d3036e9929

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

 **Discussion:**
 - delay until Sean is here

**Proposed solution or definition of done:**

**Comment by joewhitsitt** [comment]
Still want to capture a few thoughts until we can talk as a team:

- I think #9607 makes sense as long as we can temporarily patch the blt umc command or port it out so that new sites get site specific stuff still. maybe `pre-command source:build:settings` and `post-command source:build:settings` from blt/src/Blt/Plugin/Commands/ReplaceCommands.php?
- removing and replacing blt-drupal-test would allow a composer upgrade but then dealing with the other CI tooling would be good.
- at that point, do we want to consider GH Actions over Travis CI? we could disable travis for develop and see if we could get actions to work instead. if successful, we could update to handle the differences for main and tags.
- then we either need to rip the band-aid off with solving for artifact:deploy, artifact:update and the rest of ReplaceCommands or continue to chip away at the things that supplement but don't drive our work (ReportCommands, GitCommands, ConfigSplitCommands, etc.)
- I forget what the replacement for `blt ds/di/dsa` would be. Perhaps a multisite version of the commands here in .ddev/providers/acquia.yaml?
- remove blt dependencies/code
- ???
- Profit?
- D11 upgrade

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - can these be self contained?
 - audit may delay actually getting started
 - may not need to replace everything, some key critical tools though
 - need an updated approach to how we run tests, may not need to replace the blt-drupal-test command
 - replacing `umc` command could be an initial slice
   - pair with folding into https://github.com/uiowa/uiowa/pull/9607
 - validate approach in https://github.com/uiowa/uiowa/pull/9607
 - needs to account for how settings files get created locally 

**Proposed solution or definition of done:** 
- validate approach in https://github.com/uiowa/uiowa/pull/9607
- replicate functionality of `umc` command outside of BLT
  - needs options to have no commit flag, no db flag
- Provision a website sans BLT while maintaining default and site-specific settings and without disrupting flow

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - getting testing infrastructure in place to run locally
 - updating our ci process
 - shouldn't be blocked by existing in progress work
 - validating yaml in scope? unit tests?
   - separate tests from linting
   - scope to tests in this slice
 - we disabled security composer tests - would we be able to turn this back on as part of this
   - turn back on as a follow up, not as part of this issue
 - can we avoid 18 minute wait for tests to fail - order of things could be investigated
   - https://github.com/uiowa/uiowa/issues/9631#issuecomment-4209038406
 - https://dev.acquia.com/tutorial/you-dont-need-blt-acquia-cloud

**Proposed solution or definition of done:** 
- document non BLT test runner setup process
- update Travis configuration to run new test process
- optionally reorder build process to avoid long waits

Large

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - Address difficult/complex areas early (areas affecting deployment?)
 - Main legacy thing we will need to maintain is that the artifact helps keep the separate acquia repos in sync/clean
 - Epic for replacing deployment process using 30/60/90 process
   - Conceptual pass to figure out the plan to rebuild deployment process
     - due diligence + design the process
     - Could include proof of concept
     - Should include clear outline of what the replacement will look like
     - review by 1 or more developers - accepting review is approval of concept

**Proposed solution or definition of done:**
- Research and design potential new deployment process that doesn't include BLT
- Include clear outline of proposed process
  - Proof of concept acceptable but not required, implementation is out of scope for this slice
- Get feedback/approval on outlined new process from team

Investigation - Medium

**Comment by pyrello** [comment]
BLT-isms outside the #9917 deploy-process scope that still need a home in the removal:

- **Data-operation cloud hooks** — `post-db-copy`, `db-scrub`, `post-files-copy` (hook scripts call `blt artifact:ac-hooks:*`). Triggered by DB/files copy operations, not code deploys. Need non-BLT replacements when the hooks are reworked.
- **`uiowa:git:clean` (ugc)** — prunes old `{tag}-build` artifacts from the Acquia remotes. Artifact housekeeping; rides with whatever artifact model replaces `artifact:deploy`.

Surfaced while scoping #9917 (deploy-process steps tracked there).

_drafted by Claude on behalf of @pyrello_

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - replacement of deployment makes sense as a next step.
 - some blt'isms will need to stay until we switch to drupal recommended settings
 - implement as one slice of work - 30/60/90 process useful
 - frontend is standalone, Sean to verify accounted for in process

**Proposed solution or definition of done:** 
- implementation of plan from https://github.com/uiowa/uiowa/issues/9917 once feedback from team

Large

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - uiowa:multisite:execute?
 - ability to run commands against all sites
 - If it could handle escaping better this time around, I’ll buy the beers (From Joe)
 - https://github.com/jigarius/drall?
 - we may be able to accomplish this without any additional dependencies and that may be preferred based on being burnt by other dependencies recently
 - direct port vs something new
   - 30/60/90 check in

**Proposed solution or definition of done:** 

medium

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:**

Complex

 **Discussion:**
 - sync family of commands - several different ways to call `ds` including `dsa`
 - Local sync but we also have post db copy (remote to remote)
 - overlapping concerns, probably best to tackle all together
 - Scoped to include the following:
   - sync to local (ds)
   - sync local bulk (dsa)
   - remote to remote (post db copy)
 - goal would be to have parity with existing functionality
 - get rid of anything we don't need (extensions, etc. related to sync commands)
 - room for improvements if we identify easy wins
 - follow up to audit what is left, what we can get rid of
   - BLT commands that exist as fallback to new commands could be cleaned up as part of audit
   - BLT related packages that are no longer dependencies
   - custom code, traits, etc.
   - large

**Proposed solution or definition of done:**
- We can sync a single site locally using a command that is not dependent on BLT
- We can bulk sync sites locally using a command that is not dependent on BLT
- We can sync remote to remote (post db copy) using a command that is not dependent on BLT
- We have feature parity with BLT sync commands
- We have removed any extensions or other clean up related to BLT sync commands

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complicated

 **Discussion:**
 - possibly address `BLT umi` next
 - run directly on server currently to install site after created
   - could possibly be run locally
 - `BLT umd`- delete
   - we've used manual steps to delete previously
 - `BLT umt` - move sites - haven't used much, may not need this initially
 - Scope to address transition `blt umi` - sized as medium
 - could possibly be safe to remove `blt umi` as part of this

**Proposed solution or definition of done:**
- Replace `blt umi` command with a non-BLT replacement
- Shouldn't have any loss of functionality from existing `blt umi` command
- Any cleanup related to `blt umi` that we are comfortable with
